executor/infoschema_reader.go

```diff
@@ -2820,6 +2822,145 @@ func (e *memtableRetriever) setDataForRegionLabel(ctx sessionctx.Context) error
 	return nil
 }
 
+func (e *memtableRetriever) setDataFromPlacementRules(ctx context.Context, sctx sessionctx.Context, schemas []*model.DBInfo) error {
+	checker := privilege.GetPrivilegeManager(sctx)
+	is := sctx.GetInfoSchema().(infoschema.InfoSchema)
+	var rows [][]types.Datum
+
+	// Get global PLACEMENT POLICIES
+	// Currently no privileges needed for seeing global PLACEMENT POLICIES!
+	for _, policy := range is.AllPlacementPolicies() {
+		// Did we say that we would skip syntactic sugar?
+		// I proposed it to be transformed, xhebox proposed to keep it :)
+		// How about to do both, then one can see if syntactic sugar was used?
+
+		// TODO: if PrimaryRegion or Regions are set,
+		// also convert them to LeaderConstraints and FollowerConstraints
+		// for better user experience searching for particular constraints
+
+		row := types.MakeDatums(
+			policy.ID,
+			infoschema.CatalogVal, // CATALOG
+			policy.Name.O,         // Policy Name
+			nil,                   // dbName,                // SCHEMA
+			nil,                   // tbName,                // TABLE
+			nil,                   // ptName,                // PARTITION
+			policy.PlacementSettings.PrimaryRegion,
+			policy.PlacementSettings.Regions,
+			policy.PlacementSettings.Constraints,
+			policy.PlacementSettings.LeaderConstraints,
+			policy.PlacementSettings.VoterConstraints,
+			policy.PlacementSettings.FollowerConstraints,
+			policy.PlacementSettings.LearnerConstraints,
+			policy.PlacementSettings.Schedule,
+			policy.PlacementSettings.Voters,
+			policy.PlacementSettings.Followers,
+			policy.PlacementSettings.Learners,
+		)
+		rows = append(rows, row)
+	}
+
+	// Get DIRECT PLACEMENT from schemas/tables/partitions
+	for _, schema := range schemas {
+		// TODO: Filter on schema, to avoid iterating over every schema if SELECT * FROM placment_rules WHERE SCHEMA_NAME IN ('schema1', 'schema2')
```

Comment:
There's an alternative infoschema reader implementation which includes a filter for schema. See I think is.columns or is.tables for the implementation.

Reply:
I only see that infoschema.TableColumns is having a different batching approach, by implementing `(*hugeMemTableRetriever) retrieve` instead of `(*memtableRetriever) retrieve`. Maybe adding filter is a future possible optimisation?

Reply:
Yes, we can defer for later. Because the implementation is only tables with explicit placement defined, it's not a blocking issue. I would maybe change it from "TODO:" to a description of what it does, since all infoschema tables suffer from the non-pushdown behavior, and it's not specific to this table.